From 3dd63ec848c18d6136dda98522d3ccc1f000c89a Mon Sep 17 00:00:00 2001
From: Wang Wensheng <wangwensheng4@huawei.com>
Date: Tue, 19 Mar 2024 15:48:04 +0800
Subject: [PATCH 879/960] downstream: iommu/arm-smmu-v3: Disable ECMDQ before
 reset

In kdump kernel, the ECMDQ is enabled before and cannot be reset before
it is disabled. This leads to "CMD_SYNC timeout" problem.

Disable the ECMDQ before reset the ECMDQ to fix.

Signed-off-by: Wang Wensheng <wangwensheng4@huawei.com>
Signed-off-by: Deng Guangxing <dengguangxing@huawei.com>
Signed-off-by: Zhang Guangzhi <zhangguangzhi3@huawei.com>
Signed-off-by: Li Wei <liwei818@huawei.com>
---
 drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
index 1a940d617a8e..969802f327c0 100644
--- a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
+++ b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
@@ -4587,6 +4587,20 @@ static int arm_smmu_ecmdq_reset(struct arm_smmu_device *smmu)
 			q->llq.prod = 0;
 			q->llq.cons = 0;
 		}
+
+		reg = readl(q->prod_reg);
+		if (reg & ECMDQ_PROD_EN) {
+			/* disable ecmdq */
+			writel(reg & ~ECMDQ_PROD_EN, q->prod_reg);
+			ret = readl_relaxed_poll_timeout(q->cons_reg, reg,
+				!(reg & ECMDQ_CONS_ENACK), 1, ARM_SMMU_POLL_TIMEOUT_US);
+			if (ret) {
+				dev_warn(smmu->dev, "ecmdq[%d] disable failed\n", i);
+				smmu->ecmdq_enabled = 0;
+				return ret;
+			}
+		}
+
 		writeq_relaxed(q->q_base, ecmdq->base + ARM_SMMU_ECMDQ_BASE);
 		writel_relaxed(q->llq.prod, ecmdq->base + ARM_SMMU_ECMDQ_PROD);
 		writel_relaxed(q->llq.cons, ecmdq->base + ARM_SMMU_ECMDQ_CONS);
-- 
2.43.0

